synapse/handlers/auth.py

```diff
@@ -367,6 +360,41 @@ async def validate_user_via_ui_auth(
 
         return params, session_id
 
+    async def _get_available_ui_auth_types(self, user: UserID) -> Iterable[str]:
+        """Get a list of the authentication types this user can use
+        """
+
+        ui_auth_types = set()
+
+        # if the HS supports password auth, and the user has a non-null password, we
+        # support password auth
+        if self._password_localdb_enabled and self._password_enabled:
+            lookupres = await self._find_user_id_and_pwd_hash(user.to_string())
+            if lookupres:
+                _, password_hash = lookupres
+                if password_hash:
+                    ui_auth_types.add(LoginType.PASSWORD)
+
+        # also allow auth from password providers
+        for provider in self.password_providers:
+            for t in provider.get_supported_login_types().keys():
+                if t == LoginType.PASSWORD and not self._password_enabled:
+                    continue
+                ui_auth_types.add(t)
```

Comment:
👍 I believe that the Element clients used to prefer password over SSO for UI Auth, I don't know if it is due to the first available flow or what though.

Reply:
> I  believe that the Element clients used to prefer password over SSO for UI Auth, I don't know if it is due to the first available flow or what though. 

yes, they are hardcoded to do so; it's a thing that I hope they will fix ultimately, but it's not so important. Since it will only affect users who have upgraded from password-auth to SSO, hopefully those users still know their passwords...